AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift

```diff
@@ -396,13 +396,20 @@ final public class AWSMobileClient: _AWSMobileClient {
     public func showSignIn(presentationAnchor: ASPresentationAnchor,
                            hostedUIOptions: HostedUIOptions,
                            _ completionHandler: @escaping(UserState?, Error?) -> Void) {
+        if case .signedIn = self.currentUserState {
```

Comment:
The "sign in" ambiguity is throwing me. What does `.signedIn` actually mean? In other words, which of the following are known to be true if the state is `.signedIn`?

- User is identified (we have a user ID)
- User session has valid refresh token
- User session has valid access token

Depending on that answer, I'm concerned this check will require apps to call `signOut` before signing in with an expired refresh token, as noted https://github.com/aws-amplify/amplify-ios/issues/1204

Reply:
Userstate has the following states - signedIn, signedOut, signedOutFederatedTokensInvalid, signedOutUserPoolsTokenInvalid, guest, unknown.

User is identified (we have a user ID) - can be signedIn or signedOutUserPoolsTokenInvalid (when refresh token invalid)
User session has valid refresh token - signedIn
User session has valid access token - signedIn
User session has invalid refresh token - signedOutUserPoolsTokenInvalid

Reply:
| State | Valid User ID | Valid CUP Access Token | Valid CUP Refresh Token | Identity Id | AWS credentials |
| --- | --- | --- | --- | --- | --- |
| `guest` | NA | NA | NA | Yes | Yes |
| `signedIn` | Yes or NA | Yes or NA | Yes or NA | Yes | Yes |
| `signedOut` | NA | NA | NA | NA | NA |
| `signedOutFederatedTokensInvalid` | NA | NA | NA | Yes | No |
| `signedOutUserPoolsTokenInvalid` | Yes | No | No | Yes | No |
| `unknown` | ? | ? | ? | ? | ? |

Explanation
---

### `guest`
guest  status means you are using the app as an unauthenticated user. There is no user signed in to the CUP so CUP related tokens are not valid in `guest`, the only credentials available at this time are `identityID` and `aws credentials` retrieved via Cognito Identity Pool. The AWS credentials will automatically refreshed as long as you are in `guest` mode.

### `signedIn`

User can be in `signedIn` state when they are signed in to either CUP or CIDP. If they are `signedIn` to CUP they will have valid CUP access and refresh tokens. On the other hand if they are `signedIn` to CIDP, CUP related tokens are not valid, the only credentials available at this time are identityID and `aws credentials` retrieved via CIDP. 
PS: Our SDK doesnot store any social tokens.

### `signedOut`

User was signed out from the app by calling the signedOut api and the app is not configured to be in guest mode.

### `signedOutFederatedTokensInvalid `

Only `signedIn` state can transition to `signedOutFederatedTokensInvalid` if the following condition are true: i) user was in `signedIn` by signing to CIDP ii) AWS credential fetch failed because the JWT token got invalid. 

### `signedOutUserPoolsTokenInvalid `

Only `signedIn` state can transition to `signedOutUserPoolsTokenInvalid ` if the following condition are true: i) user was in `signedIn` by signing to CUP ii) AWS token fetch failed because the JWT token got invalid. 

### `unknown`

`unknown` state will only happen when the app starts or when there is some error. This is not a valid state and should never occur.

> Followup questions: How do these states relate to each of our supported authN & authZ methods? Identity Pools? Non-CUP OIDC? SAML? Custom? At what points is it valid to call a method to get credentials and tokens? 

I have added Identity Pools as a the last two columns above. For Non-CUP OIDC/SAML/Custom/ they will be signed in to the CUP, hence it will have the same properties like a user signed in with username/pwd.

You can call getToken if the status is `signedIn` and the user was signed in to CUP.
You can call getAWSCredentials if the status is either `guest` or `signedIn` and a CIDP is configured.

Reply:
Added is PR for doc site - https://github.com/aws-amplify/docs/pull/3260

Reply:
Thanks for the explanation and work putting together the docs!

The `signedOut*` states are quite confusing, since they have different behaviors for IDs & tokens depending on whether things "just expired" or whether they were invoked by calling an explicit `signOut` method. Not something to fix today, but definitely something to keep in mind for the future.

Further confusion:

> You can call getAWSCredentials if the status is either `guest` or `signedIn` and a CIDP is configured.

So if I am configured with a CUP and IP, and am in `signedOut`, I don't automatically get access to unauth AWS credentials?